src/math/exp_fcn.c

```diff
@@ -0,0 +1,153 @@
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *ptroutbitshi,
+				uint64_t *ptroutbitslo);
+/*
+ * Arguments	: int64_t a
+ *		  int64_t b
+ * Return Type	: int64_t
+ */
+static inline int64_t lomul_s64_SR_sat_near(int64_t a, int64_t b)
+{
+	uint64_t u64_rhi;
+	uint64_t u64_rlo;
+	bool roundup;
+
+	wide_mul_s64(a, b, &u64_rhi, &u64_rlo);
+	roundup = (u64_rlo & BIT_MASK_LOW_Q27P5);
+	u64_rlo = (u64_rhi << 36 | u64_rlo >> 28) + roundup;
+	return u64_rlo;
+}
+
+/*
+ * Arguments	: int64_t in_0
+ *		  int64_t in_1
+ *		  uint64_t *ptroutbitshi
+ *		  uint64_t *ptroutbitslo
+ * Multiplication of two signed int64 numbers
+ * Return Type	: void
+ */
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *ptroutbitshi,
+				uint64_t *ptroutbitslo)
+{
+	uint64_t absin0 = ABS(in_0);
+	uint64_t absin1 = ABS(in_1);
+	uint64_t in0hi = absin0 >> 32;
+	uint64_t in0lo = absin0 & UINT32_MAX;
+	uint64_t in1hi = absin1 >> 32;
+	uint64_t prodhilo;
+	uint64_t prodlohi;
+
+	absin0 = absin1 & UINT32_MAX;
+	// multiplication
+	prodhilo = in0hi * absin0;
+	prodlohi = in0lo * in1hi;
+	absin0 *= in0lo;
+	absin1 = 0;
+	in0lo = absin0 + (prodlohi << 32);
+	absin1 = in0lo < absin0 ? 1 : 0;
+	absin0 = in0lo;
+	//shift and add
+	in0lo += prodhilo << 32;
+	if (in0lo < absin0)
+		absin1++;
+
+	absin0 = absin1 + in0hi * in1hi + (prodlohi >> 32) + (prodhilo >> 32);
+	if (in_0 != 0 && in_1 != 0 && (in_0 > 0) != (in_1 > 0)) {
+		absin0 = ~absin0;
+		in0lo = ~in0lo;
+		in0lo++;
+		if (!in0lo)
+			absin0++;
+	}
+	// pointer- output high and low bytes
+	*ptroutbitshi = absin0;
+	*ptroutbitslo = in0lo;
+}
+
+/* f(x) = a^x, x is variable and a is base
+ *
+ * Arguments	: int32_t x(Q4.28)
+ * input range -5 to 5
+ *
+ * Return Type	: int32_t ts(Q9.23)
+ * output range 0.0067465305 to 148.4131488800
+ *+------------------+-----------------+--------+--------+
+ *| x		     | ts (returntype) |   x    |  ts    |
+ *+----+-----+-------+----+----+-------+--------+--------+
+ *|WLen| FLen|Signbit|WLen|FLen|Signbit| Qformat| Qformat|
+ *+----+-----+-------+----+----+-------+--------+--------+
+ *| 32 | 28  |  1    | 32 | 23 |   0   | 4.28   | 9.23   |
+ *+------------------+-----------------+--------+--------+
+ */
+int32_t exp_int32(int32_t x)
+{
+	int64_t dividend = NUM_Q14P18;
+	int32_t ts =  0; /* Q23.9 */
+	int32_t n = 0; /* Q5.26 */
+	int64_t qt;
+	uint64_t ou0Hi;
+	uint64_t ou0Lo;
+
+	// pre-computation of 1st & 2nd terms
+	ts =  TERMS_Q23P9; /* Q23.9 */
+
+	dividend = (x + LSHIFT_BITS) >> 14; //x in Q50.14
+
+	wide_mul_s64(dividend, BIT_MASK_Q62P2, &ou0Hi, &ou0Lo);
+	qt = (ou0Hi << 46) | (ou0Lo >> 18);/* Q6.26 */
+	ts += (int32_t)((qt >> 35) +
+			((qt & QUOTIENT_SCALE) >> 18));
+	dividend = lomul_s64_SR_sat_near(dividend, (int64_t)x);
+	do {
+		wide_mul_s64(dividend, iv[n], &ou0Hi, &ou0Lo);
+		qt = (int64_t)(ou0Hi << 45) | (ou0Lo >> 19);
+
+		// sum of the remaining terms
+		ts += (int32_t)((qt >> 35) +
+				((qt & QUOTIENT_SCALE) ? 1 : 0));
+		n++;
```

Comment:
Naively, putting this in a do/while loop like this means that n will eventually increment off the end of the iv[] vector and dereference garbage memory.  Now, I buy that the algorithm as designed is guaranteed to converge, so this will never happen.  But (1) that's only true until that vector gets miscopied or drops an entry and (2) there's no way static analysis tools could ever prove that, so you'll have to fix this anyway.

Suggest replacing the loop with something like: `for (int32_t n = 0; n < ARRAY_SIZE(iv) && qt <= CONVERG_ERROR; n++)`

Reply:
I think you meant `n < ARRAY_SIZE(iv) && qt> CONVERG_ERROR`.

Or maybe even better:
```C
for( ...; n < ARRAY_SIZE(iv) ;... )
{
  ...
  if (qt <= CONVERG_ERROR)
      break;
  ...
}
if (qt > CONVERG_ERROR) {
   log(some error)
   return some error
}
```

Reply:
Ah, indeed.  And feeding back to my other comment: note that almost half of the terms in iv[] are in fact 32 bit numbers, so their use in the hand-coded 64 bit multiplication in fact does hurt us: an unrolled loop would be able to use a 32x64 narrow multiplication routine for those terms instead of the full width variant.